### PR TITLE
Added thread safety to shape components

### PR DESCRIPTION
**Pull request by mbalfour-amzn**
This set of changes enables the shape components to be safely used in multithreaded environments. The shapes can be queried concurrently from multiple threads and will safely block against data-changing calls and bus connects/disconnects.

* Created unit tests that attempt to demonstrate the threading issues with the shape components, and verifies the changes are equally or more threadsafe. The unit tests only reliably demonstrated threading errors on 4/10 of the shape components (Capsule, Compound, Polygon Prism, Reference) prior to the fixes, I couldn't find a way to make them reliably fail on the other shape components, as their data usage was so minimal that either the data was never inconsistent mid-operation or else inconsistent data didn't end up affecting the results, depending on the shape. Regardless, even for those shapes these tests help to demonstrate that the changes continue to work at least as well as before.
* Added mutexes to allow for concurrent queries while blocking on data changes
* Modified shape request bus to allow for concurrent queries while blocking on bus connects / disconnects

Also found/fixed an unrelated bug where the Polygon Prism Shape component was keeping stale callback pointers around and triggering them as a part of the Activate() sequence before the component had fully initialized itself or the callbacks.

### CODE REVIEW

File: Gems/LmbrCentral/Code/Source/Shape/BoxShape.cpp [hunk 1]
```diff
@@ -79,39 +79,53 @@ namespace LmbrCentral
...
         ShapeComponentNotificationsBus::Event(
             m_entityId, &ShapeComponentNotificationsBus::Events::OnShapeChanged,
             ShapeComponentNotifications::ShapeChangeReasons::ShapeChanged);
     }
 
     AZ::Aabb BoxShape::GetEncompassingAabb()
     {
-        m_intersectionDataCache.UpdateIntersectionParams(m_currentTransform, m_boxShapeConfig, m_currentNonUniformScale);
+        AZStd::shared_lock lock(m_mutex);
+        m_intersectionDataCache.UpdateIntersectionParams(m_currentTransform, m_boxShapeConfig, m_mutex, m_currentNonUniformScale);
```
Comment (lumberyard-employee-dm) [inline Gems/LmbrCentral/Code/Source/Shape/BoxShape.cpp:127]: Is the `m_intersectionDataCache` being modified by the `UpdateIntersectionParams` function or is it a const member?

File: Gems/LmbrCentral/Code/Source/Shape/PolygonPrismShape.h [hunk 2]
```diff
@@ -103,12 +104,78 @@ namespace LmbrCentral
...
+        // callbacks to query the PolygonPrismShape to refresh some data.
+        // This means that when modifying data, the PolygonPrismShape can lock a unique_lock, but while it's holding that lock, callbacks
+        // can trigger that try to read data on the same thread, which will try to lock a shared_lock.
+        // The extra logic the PolygonPrism*LockGuard classes perform is that they track the thread that acquires a unique lock,
+        // and they don't try to acquire or release a shared lock if one is requested on the same thread as the unique lock.
+
+        class PolygonPrismSharedLockGuard
+        {
+        public:
+            explicit PolygonPrismSharedLockGuard(AZStd::shared_mutex& mutex, AZStd::thread::id& uniqueLockThreadId)
```
Comment (lumberyard-employee-dm) [inline Gems/LmbrCentral/Code/Source/Shape/PolygonPrismShape.h:121]: explicit is not need for a function that accepts two required parameters.

File: Gems/LmbrCentral/Code/include/LmbrCentral/Shape/ShapeComponentBus.h [hunk 3]
```diff
@@ -105,11 +114,11 @@ namespace LmbrCentral
     };
 
     /// Services provided by the Shape Component
-    class ShapeComponentRequests : public AZ::ComponentBus
+    class ShapeComponentRequests : public AZ::EBusSharedDispatchTraits<ShapeComponentRequests>
     {
     public:
-        /// allows multiple threads to call shape requests
-        using MutexType = AZStd::recursive_mutex;
+        static const AZ::EBusAddressPolicy AddressPolicy = AZ::EBusAddressPolicy::ById;
```
Comment (lumberyard-employee-dm) [inline Gems/LmbrCentral/Code/include/LmbrCentral/Shape/ShapeComponentBus.h:120]: nitpick: You should make this `constexpr`

File: Gems/LmbrCentral/Code/Source/Shape/BoxShape.cpp (hunk 1 above)
Comment (mbalfour-amzn) [reply Gems/LmbrCentral/Code/Source/Shape/BoxShape.cpp:127]: UpdateIntersectionParams will check to see if the cache needs to be updated, and then update it if it does. To avoid needing unique locks in the common case where the cache is up-to-date, this passes in the mutex, and UpdateIntersectionParams() will change it from a shared_lock to a unique_lock while updating the cache, and then back to a shared_lock before exiting.

File: Gems/LmbrCentral/Code/Source/Shape/PolygonPrismShape.h (hunk 2 above)
Comment (mbalfour-amzn) [reply Gems/LmbrCentral/Code/Source/Shape/PolygonPrismShape.h]: Fixed!

File: Gems/LmbrCentral/Code/include/LmbrCentral/Shape/ShapeComponentBus.h (hunk 3 above)
Comment (mbalfour-amzn) [reply Gems/LmbrCentral/Code/include/LmbrCentral/Shape/ShapeComponentBus.h]: Fixed!

**Review by cgalvan — APPROVED** [review]

File: Gems/LmbrCentral/Code/Tests/ShapeThreadsafeTest.cpp [hunk 4]
```diff
@@ -43,7 +43,7 @@ namespace UnitTest
         for (auto& queryThread : queryThreads)
         {
             queryThread = AZStd::thread(
-                [shapeEntityId, TestPoint, numIterations, &syncThreads]()
+                [shapeEntityId, TestPoint, ExpectedDistance, numIterations, &syncThreads]()
```
Comment (lumberyard-employee-dm) [inline Gems/LmbrCentral/Code/Tests/ShapeThreadsafeTest.cpp:46]: If you want something that works across clang and msvc you can use a named capture  `expectedDistance = ExpectedDistance`.

File: Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp [hunk 5]
```diff
@@ -164,4 +165,31 @@ namespace UnitTest
...
+        {
+            float x = minDimension + aznumeric_cast<float>(rand() % dimensionVariance);
+            float y = minDimension + aznumeric_cast<float>(rand() % dimensionVariance);
+
+            LmbrCentral::BoxShapeComponentRequestsBus::Event(
+                shapeEntityId, &LmbrCentral::BoxShapeComponentRequestsBus::Events::SetBoxDimensions, AZ::Vector3(x, y, height));
+        };
+
+        // Run the test, which will run multiple queries in parallel with each other and with the dimension-setting function.
+        const int numIterations = 100000;
```
Comment (lumberyard-employee-dm) [inline Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp:191]: Do we really need a 100000 iterations?
I wonder about the trade-off between trying to detect a race condition vs test time
Any thoughts @Kadino ?

File: Gems/LmbrCentral/Code/include/LmbrCentral/Shape/ShapeComponentBus.h [hunk 6]
```diff
@@ -105,11 +114,11 @@ namespace LmbrCentral
...
 
     /// Services provided by the Shape Component
-    class ShapeComponentRequests : public AZ::ComponentBus
+    class ShapeComponentRequests : public AZ::EBusSharedDispatchTraits<ShapeComponentRequests>
     {
     public:
-        /// allows multiple threads to call shape requests
-        using MutexType = AZStd::recursive_mutex;
+        static constexpr AZ::EBusAddressPolicy AddressPolicy = AZ::EBusAddressPolicy::ById;
+        typedef AZ::EntityId BusIdType;
```
Comment (lumberyard-employee-dm) [inline Gems/LmbrCentral/Code/include/LmbrCentral/Shape/ShapeComponentBus.h:121]: We prefer `using` over `typedef`.

File: Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp (hunk 5 above)
Comment (Kadino) [reply Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp:191]: Agreed that this seems like an intense amount of executions. This type of race-condition fuzz-testing thrash is a bit of a paranoid way to test, so I would suggest limiting it to at most 2x the minimal set of iterations that load up the parallel threads and create the expected scenario. Better yet is to add specific tests for edge cases when they are found.

How long does the test currently take? I'd recommend targeting something lower than a full second, given how many cycles should execute. Given that it will get run in Jenkins many dozens of times per day, I expect a handful of iterations should be more than enough.

Adding a CLI parameter might reduce investigation time by a few minutes of compilation. Complicates the code a bit for a "maybe useful someday" scenario. I don't find that too compelling.

Comment (mbalfour-amzn) [reply Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp:191]: **Short version:** 
Each added test takes < 500 ms right now. If this needs to be reduced further, let me know.

**Long version:**
"Do we really need a 100000 iterations? I wonder about the trade-off between trying to detect a race condition vs test time."
I picked 100,000 as a balance between as many iterations as possible and keeping the tests well under 1 second each. Here's how much time the tests take in a profile build on my PC:
```
QuadShapeTest.ShapeHasThreadsafeGetSetCalls (119 ms)
TubeShapeTest.ShapeHasThreadsafeGetSetCalls (88 ms)
DiskShapeTest.ShapeHasThreadsafeGetSetCalls (118 ms)
ReferenceComponentTests.ShapeHasThreadsafeGetSetCalls (226 ms)
AxisAlignedBoxShapeTest.ShapeHasThreadsafeGetSetCalls (116 ms)
BoxShapeTest.ShapeHasThreadsafeGetSetCalls (86 ms)
SphereShapeTest.ShapeHasThreadsafeGetSetCalls (96 ms)
CylinderShapeTest.ShapeHasThreadsafeGetSetCalls (95 ms)
CapsuleShapeTest.ShapeHasThreadsafeGetSetCalls (111 ms)
PolygonPrismShapeTest.ShapeHasThreadsafeGetSetCalls (488 ms)
```

Just for completeness, here are the numbers from a debug build. The only egregious one (IMO) is the PolygonPrism, which could probably benefit from some optimization in the DistanceFromPointSquared method:
```
QuadShapeTest.ShapeHasThreadsafeGetSetCalls (174 ms)
TubeShapeTest.ShapeHasThreadsafeGetSetCalls (427 ms)
DiskShapeTest.ShapeHasThreadsafeGetSetCalls (160 ms)
ReferenceComponentTests.ShapeHasThreadsafeGetSetCalls (776 ms)
AxisAlignedBoxShapeTest.ShapeHasThreadsafeGetSetCalls (209 ms)
BoxShapeTest.ShapeHasThreadsafeGetSetCalls (238 ms)
SphereShapeTest.ShapeHasThreadsafeGetSetCalls (110 ms)
CylinderShapeTest.ShapeHasThreadsafeGetSetCalls (234 ms)
CapsuleShapeTest.ShapeHasThreadsafeGetSetCalls (192 ms)
PolygonPrismShapeTest.ShapeHasThreadsafeGetSetCalls (2594 ms)
```

I'm happy to drop the iteration value further if this seems overly burdensome. If so, please recommend the time target to aim for.

"This type of race-condition fuzz-testing thrash is a bit of a paranoid way to test, so I would suggest limiting it to at most 2x the minimal set of iterations that load up the parallel threads and create the expected scenario. Better yet is to add specific tests for edge cases when they are found."

To be clear, this is my absolutely _least_ favorite way to test for race conditions. I _much_ prefer setting up lockstep-style tests. However, for the shape code, I couldn't find a better way to validate the existence and removal of the race conditions that I could clearly see existing through manual code inspection. Specifically, what this is detecting is the brokenness that can occur when (for example) `CapsuleShape::CapsuleIntersectionDataCache::UpdateIntersectionParamsImpl` is updating the 6 cached values it has while `CapsuleShape::SetRadius` is changing the configured radius in-between different calculations in the update that use the radius. The race condition failure occurs when the updated cached values weren't all calculated with the same configuration values, which produces a result that doesn't match either the before _or_ the after state of the capsule shape change. This condition only occurs in the capsule case when the radius changes in-between those few lines of code.

This test may seem paranoid, but it successfully detected the race condition 100% of the time on 4/10 of the shapes prior to the fixes, and 0% of the time after the fixes. No other automated tests were detecting the problem before this was written, and this is a demonstrable problem, so I believe it to be a valid and useful test for the specific bug it is detecting. I did attempt to tune it to minimize the overall runtime while maximizing the chance of successful detection. My goal was to tune this to 100% detection as best as I could, because _intermittent_ failures on AR runs due to thrash-based unit tests is never enjoyable.

It's unfortunately also worth noting that this unit test did _not_ successfully detect the race condition on 6/10 of the shapes even prior to my fixes, mostly because their logic and number of cached values are simple enough that the changes are staying _effectively_ atomic, even if the code has no mechanisms to enforce that. I still left those tests in as well though, because they at least establish a baseline to ensure that changes to the shape code don't start causing this condition to fail.

Comment (Kadino) [reply Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp:191]: I'm interested in the threshold where the detected cases stop being reliably detected (on say 1/10 runs) on your local environment. If that currently occurs around 50k iterations, then this is a reasonable test.

Part of me would love to be a purist and say "no, never fuzz!" but the more important testing principle is to examine the pain of this regressing compared to the cost of running these tests (a few seconds per module currently translates to a few machine hours per month).

Another approach for reducing the cost, particularly for regression-searching (vs. correctness enforcing) tests, is to move tests into the periodic suite. However that entails failing changes promoting into development before later catching the issue, hopefully before they promote toward release.

File: Gems/LmbrCentral/Code/include/LmbrCentral/Shape/ShapeComponentBus.h (hunk 6 above)
Comment (mbalfour-amzn) [reply Gems/LmbrCentral/Code/include/LmbrCentral/Shape/ShapeComponentBus.h]: Fixed!

File: Gems/LmbrCentral/Code/Tests/ShapeThreadsafeTest.cpp (hunk 4 above)
Comment (mbalfour-amzn) [reply Gems/LmbrCentral/Code/Tests/ShapeThreadsafeTest.cpp]: Fixed!

File: Gems/LmbrCentral/Code/Tests/TubeShapeTest.cpp [hunk 7]
```diff
@@ -412,4 +413,37 @@ namespace UnitTest
...
+            // Set the end radii to random values. As long as we don't set the middle radii that describe the portion of the tube
+            // that intersects the origin, changing these should have no effect on our test point distance.
+
+            float radius = minDimension + aznumeric_cast<float>(rand() % dimensionVariance);
+            LmbrCentral::TubeShapeComponentRequestsBus::Event(
+                shapeEntityId, &LmbrCentral::TubeShapeComponentRequestsBus::Events::SetVariableRadius, 0, radius);
+
+            radius = minDimension + aznumeric_cast<float>(rand() % dimensionVariance);
+            LmbrCentral::TubeShapeComponentRequestsBus::Event(
+                shapeEntityId, &LmbrCentral::TubeShapeComponentRequestsBus::Events::SetVariableRadius, 3, radius);
```
Comment (lumberyard-employee-dm) [inline Gems/LmbrCentral/Code/Tests/TubeShapeTest.cpp:442]: You can pass in another lambda to the TubeShapeComponentRequestsBus so that the only 1 EBus event occurs instead of 3. That would improve performance of the test.
```
auto SetRadii = [](LmbrCentral::TubeSharedComponentRequests* requests)
{
    requests->SetRadius(height / 2.0f)
    float radius = minDimension + aznumeric_cast<float>(rand() % dimensionVariance);
    requests->SetVariableRadius, 0, radius);
    radius = minDimension + aznumeric_cast<float>(rand() % dimensionVariance);
    requests->SetVariableRadius, 0, radius);
};

LmbrCentral::TubeShapeComponentRequestsBus::Event(shapeEntityId, SetRadii);
```

**Review by lumberyard-employee-dm — APPROVED** [review]

File: Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp (hunk 5 above)
Comment (mbalfour-amzn) [reply Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp:191]: "I'm interested in the threshold where the detected cases stop being reliably detected (on say 1/10 runs) on your local environment. If that currently occurs around 50k iterations, then this is a reasonable test."
Kind of a tricky question to answer. If I run the unit tests without the shape/bus changes, I start to get < 100% reliability on Reference Component at 90 iterations, Capsule at 1000 iterations. Once I added multi-threading into the shape bus, these numbers went up to closer to the 10,000 - 15,000 range. Based solely on those numbers on my PC, we could _probably_ reduce all the iterations down to 10,000. However, if we tune the number of iterations to a specific CPU and a specific set of conditions that are now fixed, there's no particular reason to believe it is well-tuned to catch other sets of conditions that produce this result, or on other CPUs. To your point about "pain of regression", to me a few machine hours a month are an acceptable cost vs the cost of finding and fixing a regression if these components stop being threadsafe in some specific circumstance.

"Another approach for reducing the cost, particularly for regression-searching (vs. correctness enforcing) tests, is to move tests into the periodic suite."
Absolutely a fair suggestion. I'd prefer to keep these tests in the main suite due to the value I've already gotten from them though. Along with proving out the initial race condition, these unit tests also caught mistakes I made during implementation of this PR, including deadlocks that I accidentally introduced at one point due to some unexpected nested locks that could occur. Because they caught those problems, I feel like they fall more on the side of "correctness enforcing" for thread safety rather than just pure regression checking.

File: Gems/LmbrCentral/Code/Tests/TubeShapeTest.cpp (hunk 7 above)
Comment (mbalfour-amzn) [reply Gems/LmbrCentral/Code/Tests/TubeShapeTest.cpp:442]: I'd rather leave this one as separate EBus calls, as it will do a better job of spamming the EBus mutex with smaller lock/unlock cycles. 

Also, FWIW, this change makes the test performance worse - across 10 runs the average time went from 64.1 ms to 79.5 ms. The way the test works is that it runs the _query_ threads for a given number of iterations, and runs the Set thread as many times as necessary until the queries are done. By having the Set thread hold a single lock for longer, all 4 query threads are slowed down by a small amount on each iteration.

File: Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp (hunk 5 above)
Comment (Kadino) [reply Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp:191]: My suggestion above is to use double that minimal threshold of where the local repro starts to become reliable, as a heuristic of where "good enough" lies. I'm comfortable with these tests living in the Main suite, but prefer that decision stay the hands of test-authors :)

Comment (mbalfour-amzn) [reply Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp]: Sounds like a great heuristic! :) Reduced iterations on these tests to 30,000, which is 2X the 15,000 value that was the maximum needed to reliably reproduce the errors.

File: Gems/LmbrCentral/Code/Tests/TubeShapeTest.cpp (hunk 7 above)
Comment (lumberyard-employee-dm) [reply Gems/LmbrCentral/Code/Tests/TubeShapeTest.cpp:442]: The performance shouldn't be worse.
Was this logic added inside of the `SetDimensionFn` function?

Comment (mbalfour-amzn) [reply Gems/LmbrCentral/Code/Tests/TubeShapeTest.cpp:442]: Yes. Here is the specific test that I measured vs what is in the PR:
```
    TEST_F(TubeShapeTest, ShapeHasThreadsafeGetSetCalls)
    {
        // Verify that setting values from one thread and querying values from multiple other threads in parallel produces
        // correct, consistent results.

        // Create our sphere centered at 0 with half our height as the radius.
        AZ::Entity entity;
        CreateTube(AZ::Transform::CreateTranslation(AZ::Vector3::CreateZero()), ShapeThreadsafeTest::ShapeHeight / 2.0f, entity);

        // Define the function for setting unimportant dimensions on the shape while queries take place.
        auto setDimensionFn = [](AZ::EntityId shapeEntityId, float minDimension, uint32_t dimensionVariance, float height)
        {
            auto SetRadii = [minDimension, dimensionVariance, height](LmbrCentral::TubeShapeComponentRequests* requests)
            {
                requests->SetRadius(height / 2.0f);
                float radius = minDimension + aznumeric_cast<float>(rand() % dimensionVariance);
                requests->SetVariableRadius(0, radius);
                radius = minDimension + aznumeric_cast<float>(rand() % dimensionVariance);
                requests->SetVariableRadius(3, radius);
            };

            LmbrCentral::TubeShapeComponentRequestsBus::Event(shapeEntityId, SetRadii);
        };

        // Run the test, which will run multiple queries in parallel with each other and with the dimension-setting function.
        // The number of iterations is arbitrary - it's set high enough to catch most failures, but low enough to keep the test
        // time to a minimum.
        const int numIterations = 30000;
        ShapeThreadsafeTest::TestShapeGetSetCallsAreThreadsafe(entity, numIterations, setDimensionFn);
    }
```
I ran this with a gtest_repeat=10. The new averages with 30000 iterations are 28.4 ms for what's in the PR vs 31.4 ms for this version.

There isn't an easy way to capture a picture of what's happening here (I think I'd have to add the Profiler Gem to the LmbrCentral tests to show this in PIX), but my assumption for why it's slower is that the Set thread is successfully stealing more time from the query threads, making them take longer overall to finish their given number of iterations.

Basically, we've taken each loop of the Set thread from looking like this:
|----| |---| |----|
to this:
|-----------|

There are less "gaps" where the query threads have an ability to acquire their mutex, so the Set thread can potentially block all 4 query threads for longer stretches of time on each of its iterations. 

I admit I don't fully understand the threading nuances going on here though. Given that picture, I would've expected to find that the Set thread was iterating _more_ frequently with the above code change and starving out the query threads, but in fact it's executing _less_ iterations. The Set loop is executing on average 8034 iterations in the PR code, vs 7029 iterations in the code above. So it's running less times, but it must also manage to block the query threads for longer when it does run, because the query thread execution length is what dictates the total runtime length. 

I agree that this is a non-intuitive result, and in _general_ I believe in grouping EBus calls together like this, but for this particular test it doesn't appear to be a good idea.

**Review by Kadino — APPROVED** [review]
